File: day10.rb
################################################################################
# Functions
################################################################################
def transform(number_string)
  res = ""
  index = 1
  count = 1
  cur_number = number_string[0]

  while index < number_string.length
    if number_string[index] == cur_number
      #current number equals the previous one
      count += 1

    else
      #handle previous sequence
      res += count.to_s
      res += cur_number.to_s

      #start new sequence
      count = 1
      cur_number = number_string[index]
    end

    index += 1
  end


  #handle last sequence
  res += count.to_s
  res += cur_number.to_s

  return res
end

################################################################################
# Script
################################################################################

puzzle = "1"

20.times do |n|
  puzzle = transform(puzzle)
  puts "step #{n}: #{puzzle}"
end

puts "The output is #{puzzle}"

File: day10.rs
fn main() {
    let mut sequence: Vec<i32> = vec![3,1,1,3,3,2,2,1,1,3];

    for _i in 0..50 {
        let mut tmp: Vec<i32> = Vec::new();
        let mut cur_number = sequence[0];
        let mut count = 0;

        for n in &sequence {
            if cur_number == *n {
                count += 1;
            } else {
                tmp.push(count);
                tmp.push(cur_number);
                count = 1;
                cur_number = *n;
            }
        }
        tmp.push(count);
        tmp.push(cur_number);

        sequence = tmp;
    }


    println!("result size: {}", sequence.len());
}
